File: node_depths/node_depths.go
package main

import "fmt"

type BST struct {
    value int
    left  *BST
    right *BST
}

func nodeDepths(root *BST, depth int) int {
    if root == nil {
        return 0
    }
    return depth + nodeDepths(root.left, depth+1) + nodeDepths(root.right, depth+1)
}

func main() {
    bst := &BST{value: 10}
    bst.left = &BST{value: 8}
    bst.right = &BST{value: 12}

    fmt.Println(nodeDepths(bst, 0))  // Output: 4
}

File: node_depths/node_depths.java
class BST {
    int value;
    BST left;
    BST right;

     BST(int value) {
        this.value = value;
        this.left = null;
        this.right = null;
    }

    public static int nodeDepths(BST root) {
        return nodeDepthsHelper(root, 0);
    }

    private static int nodeDepthsHelper(BST node, int depth) {
        if (node == null) {
            return 0;
        }
        return depth + nodeDepthsHelper(node.left, depth + 1) + nodeDepthsHelper(node.right, depth + 1);
    }

    public static void main(String[] args) {
        BST bst = new BST(10);
        bst.left = new BST(8);
        bst.right = new BST(12);
        System.out.println(nodeDepths(bst)); 
    }
}
